Flush GSD buffers on exit.

joaander opened the pull request:
<!-- Please confirm that your work is based on the correct branch. -->
<!-- Bug fixes should be based on *trunk-patch*. -->
<!-- Backwards compatible new features should be based on *trunk-minor*. -->
<!-- Incompatible API changes should be based on *trunk-major*. -->

## Description

<!-- Describe your changes in detail. -->
Maintain a list of active GSD writers. Use an `atexit` method to flush them on exit. Also flush the files in the exception handler to attempt to keep data before calling MPI_Abort.

## Motivation and context

<!--- Why is this change required? What problem does it solve? -->
Due to #1457 or a related issue, GSD files may not flush **even when the program cleanly exits**. Explicit is better than implicit, so forcibly flushing the GSD files at exit ensures as much data as possible is written. 

Note: We may be able to solve #1457 using some `weakref` objects to break cycles.

## How has this been tested?

<!--- Please describe in detail how you tested your changes. -->
Existing unit tests pass. Examined the length of `_open_gsd_writers` with and without the `finalizer` object and got the expected results.

<!--- Please build the sphinx documentation and check that any changes to
      documentation display properly. -->

## Change log

Not needed. Fixing a new feature in this release.

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

joaander committed:
Flush GSD buffers on exit.

b-butler commented:
@joaander, Yeah I think switching all our simulation references to weakrefs could work. Or we could remove the simulation storage in many of them and optionally store a weak reference through the use of `weakref` and `property`.

joaander committed:
Flush on finalize.

This code was working a few minutes ago, then it stopped flushing.
Somehow the garbage collector started collecting the Python GSD object
but not the C++ GSDDumpWriter. Ensure that flush is called in this case
as well.

tommy-waltmann commented on hoomd/write/gsd.py:
```diff
@@ -203,6 +225,12 @@ def _attach_hook(self):
+        # Maintain a list of open gsd writers
+        weak_writer = weakref.ref(self)
+        _open_gsd_writers.append(weak_writer)
+        self._finalizer = weakref.finalize(self, _finalize_gsd, weak_writer,
+                                           self._cpp_obj),
```
It is possible to do this in `__del__` instead of using `weakref.finalize`?

joaander replied:
Yes. However the finalizer is more flexible for future development. For example if we were to add a `close` method to `GSD`, the finalizer is a cleaner solution. See this example for an explanation: https://docs.python.org/3/library/weakref.html#comparing-finalizers-with-del-methods

tommy-waltmann reviewed (commented):
There are a couple of other potential solutions I'd like to discuss. I mentioned one in a comment and I'll mention another below:

Can we just call `flush` from the destructor of the `GSDDumpWriter`? That destructor is guaranteed to run in every case.

b-butler commented:
@tommy-waltmann Actually, since the process is a Python interpreter, the destructor of any object is not guaranteed to be called if Python just decides to exit without cleaning up all its memory (which it does) this is why we cannot use `__del__` or destructors, because when the kernel cleans up Python's mess, it has no idea what byte represents what.

joaander commented:
The destructor of `GSDDumpWriter` implicitly calls flush via `gsd_close()`. However, this was not being called in `hoomd-validation`'s `lj-fluid` project, *even for clean exits*.

These are the five behaviors I have seen or been able to create and how they are handled by this pull request.
1. Clean exit / SIGTERM (or unhandled exception in serial): Python's garbage collector cleans up both the Python level `GSD` object and the C++ `GSDDumpWriter`
  *  The finalizer triggers, flushes the file and removes it from `_open_gsd_writers`.
  * `GSDDumpWriter` closes the file.
  * `_open_gsd_writers` is empty and the `atexit` routine does nothing.
2. Clean exit / SIGTERM (or unhandled exception in serial): Python's garbage collector cleans up the Python level `GSD` object *but not* the C++ `GSDDumpWriter`
  *  The finalizer triggers, flushes the file and removes it from `_open_gsd_writers`.
  * `_open_gsd_writers` is empty and the `atexit` routine does nothing.
  * The files are closed by the kernel during tear down.
3. Clean exit / SIGTERM (or unhandled exception in serial): Python's garbage collector fails to clean up both the Python level `GSD` object and the C++ `GSDDumpWriter`
  * `_open_gsd_writers` contains the writers and the `atexit` routine flushes the files.
  * The files are closed by the kernel during tear down.
4. Unhandled exception in MPI on the root rank.
  * `_hoomd_sys_excepthook` flushes the write buffers in all writers in `_open_gsd_writers`, then calls `MPI_Abort` which kills the other rank's processes.
  * The files are closed by the kernel during tear down.
5. Unhandled exception in MPI on a non-root rank.
  * `_hoomd_sys_excepthook` calls `MPI_Abort` which kills the other rank's processes including the root rank. This is `SIGKILL` which cannot be caught, so buffers are not flushed.
  * The files are closed by the kernel during tear down.
 
Unhandled exceptions thrown on all ranks can potentially follow either 4 or 5 depending on race conditions (I got 4 in a few quick tests). There is no way to avoid this without causing deadlock. I add the call to hopefully preserve intended GSD file contents as much as possible.

Behaviors 2 and 3 do not appear in minimal examples. #1457 shows one. The `lj-fluid.py` I am currently working on is another. That one gave behavior 3 reproducibly for most of the morning. I took a break and went back later in the day, then it gave behavior 2 reproducibly. You can test this by running with `notice_level=5`. If the garbage collector works correctly, you should see many object deconstruction notices and notices that the GSD file are closed.

Even if we could somehow work around the bugs in Python's garbage collector so that behaviors 2 and 3 never occur, we will need this PR to cover 4 even if it is not perfect.

joaander commented:
Due to the way that `MPI_Abort` works, uncaught exceptions in one partition will send `SIGKILL` to the other partitions (at least in OpenMPI): https://www.open-mpi.org/doc/v4.1/man3/MPI_Abort.3.php
> The long-term goal of the Open MPI implementation is to terminate all processes in all tasks that contain a process in comm, and the error code is not returned to the invoking environment. At the moment, this isn’t fully implemented and MPI_Abort will terminate the entire job.

This means that users wishing to obtain partial data files from jobs that fail should either catch exceptions and/or explicitly call `flush` at specific points needed to ensure data integrity (i.e. flush trajectory files when writing a restart file). Perhaps we should demonstrate this in the workflow tutorial.

tommy-waltmann commented:
It looks like there are a lot of details related to cleaning up that I had not anticipated. It seems that this PR handles the possible outcomes as cleanly as it can (outcome 5 doesn't seem to be fixable).
